### ISSUE TITLE
A Certificate is required to complete the authentication

### ISSUE THREAD

**Issue by raghavmp79** [libgit2/libgit2sharp#1152]
Hi,

When I trying to clone a repository, I am getting an Unhandled error message.
"send request failed: A certificate is required to complete the authentication"
Please let me know what wrong I am doing.

CloneOptions co = new CloneOptions();
            co.CredentialsProvider = (_url, _user, _cred) => new UsernamePasswordCredentials { Username = "abc", Password = "abc@123" };

            Repository.Clone("https://abc.com/stash/scm/work.git", @"D:\EP", co);

public static RepositorySafeHandle git_clone(
            string url,
            string workdir,
            ref GitCloneOptions opts)
        {
            RepositorySafeHandle repo;
            int res = NativeMethods.git_clone(out repo, url, workdir, ref opts);
            **Ensure.ZeroResult(res);**
            return repo;
        }
Error line is in bold

**Comment by carlosmn** [comment]
That message seems to come from WinHTTP, likely indicating that the server wanted a certificate from the client (though the usual message for that is somewhat different from the one you quote above). If you do need a client certificate, then there is nothing you can do with the current libgit2sharp code.

How do you normally connect to the server with git? Do you use a client certificate? It's rather unusual that you'd specify a username/password and then be asked for a certificate? Does it work with git? How do you set it up?

**Comment by raghavmp79** [comment]
Yes! we use client certificate to connect the server. We provide username/password and then server asks for client certificate. It works fine with the browser.

**Comment by carlosmn** [comment]
The majors browsers do support client certs, but they don't speak Git. This is something which is just not supported by libgit2(sharp). WinHTTP should support them, so it should be theoretically possible to add this functionality on Windows, but it is rather an edge case of functionality.

**Comment by krystan** [comment]
I just spoke to the guys over at axosoft (gitkraken). I was trying to pull from tfs 2017. And was unable to and got this same error. Gitkraken takes an indirect dependency on this project through nodegit. Is this still considered edge functionality because if so microsofts pretty standard config will not work with any project taking a dependency on this library.

**Comment by ethomson** [comment]
Hi @krystan .  GitKraken uses nodegit which is a wrapper around [_libgit2_](github.com/libgit2/libgit2); it's not a wrapper around this project, LibGit2Sharp, which itself is the .NET wrapper around libgit2.

libgit2 does not support client certificates at this moment, though I would recommend creating an issue on the libgit2 project if that's something you're interested in.

**Comment by whoisj** [comment]
>  I was trying to pull from tfs 2017.

TFS 2017 supports Personal Access Tokens, specifically to resolve these kinds of issue.

https://www.visualstudio.com/en-us/docs/setup-admin/team-services/use-personal-access-tokens-to-authenticate

**Comment by etomm** [comment]
I have the same problem with BitBucket local instance

**Comment by cguanes** [comment]
Something here with Bitbucket hosted locally version 5.6, we use a certificate created on Windows server certificate. We browser to bitbucket works fine on https, Sourcetree too.

**Comment by ethomson** [comment]
> Something here with Bitbucket hosted locally version 5.6, we use a certificate created on Windows server certificate. Web browser to bitbucket works fine on https, Sourcetree too.

That's a _server certificate_.  This issue is a feature request about supporting the ability to authenticate to the server using a client certificate.  If you have problems accepting a server certificate, please make sure that your certificate is _signed_, or you have otherwise accepted the server's certificate with the system certificate store, _or_ you have accepted the certificate with the callback.

If you have trouble, please open an issue on [StackOverflow](https://stackoverflow.com/questions/tagged/libgit2sharp) to get support.
